### PR TITLE
Rework article on ADTs

### PR DESCRIPTION
**Pull request by tindzk**

### CODE REVIEW

File: articles/2016-10-12-functor-adts.txt [hunk 1]
```diff
@@ -1,37 +1,38 @@
 package[value="articles"]{}
 
-post[id=functor-adts,date=2016-10-12,description="Using contravariance, type parameters and type members to implement a functor in ADTs",Implementing functors in ADTs] {
+post[id=functor-adts,date=2016-10-12,description="Review of Scala language features for call site type polymorphism",Call site type polymorphism] {
   section[Motivation] {
-    It is a common problem to define url[https://en.wikipedia.org/wiki/Algebraic_data_type]{Algebraic Data Types} (ADTs). To prevent duplication, shared functionality should be ideally kept in base code{trait}s. Oftentimes, a problem arises that in the base code{trait} we do not know the child type the function is called on.
+    It is a common problem to define class hierarchies. To prevent duplication, shared functionality should be ideally kept in base code{trait}s. Oftentimes, a problem arises that in the base code{trait} we do not know the child type the function is called on:
```
Comment (anatoliykmetyuk) [inline articles/2016-10-12-functor-adts.txt]: We do not know the type the child method returns, not the type the function is called on. If they are the same, `this.type` is the shortest solution.

File: articles/2016-10-12-functor-adts.txt [hunk 2]
```diff
@@ -1,37 +1,38 @@
...
 
-    In other words, we want to be able to implement operations such as code{map()} in terms of a functor:
-    scala[type=section,file=FunctorAdts,value=functor]{}
+    scala[type=section,file=FunctorAdts,value=problem]{}
 
-    In this article, I am going to outline different ways how to solve this particular problem in Scala.
+    In other words, for a class code{F[_]} we want to define methods in its base code{trait}. These methods should have the return type code{F[_]}. In this article, I am going to outline different ways how to solve this particular problem in Scala.
```
Comment (anatoliykmetyuk) [inline articles/2016-10-12-functor-adts.txt]: Why `F[_]` and not `F`?

File: articles/2016-10-12-functor-adts.txt [hunk 3]
```diff
@@ -1,37 +1,38 @@
...
   }
 
   section[Contravariance] {
     Consider a simple binary tree:
     scala[type=section,file=FunctorAdts,value=bin-tree]{}
 
-    We are going to leverage the fact that functions are contravariant in their return types. Let us extend our ADT by a code{map()} function:
+    We are going to leverage the fact that functions are contravariant in their return types. Let us extend our hierarchy by a code{map()} function:
```
Comment (anatoliykmetyuk) [inline articles/2016-10-12-functor-adts.txt]: The functions are co-variant in their return types and contra-variant in their input types: http://www.scala-lang.org/api/current/index.html#scala.Function1

File: articles/2016-10-12-functor-adts.txt (hunk 1 above)
Comment (tindzk) [reply articles/2016-10-12-functor-adts.txt]: As per you suggested, I added a section on `this.type`. How would you rephrase this paragraph to make it more clear?

File: articles/2016-10-12-functor-adts.txt (hunk 2 above)
Comment (tindzk) [reply articles/2016-10-12-functor-adts.txt]: Indeed, the type parameter is redundant.

File: articles/2016-10-12-functor-adts.txt (hunk 3 above)
Comment (tindzk) [reply articles/2016-10-12-functor-adts.txt]: You're right. Thanks!

File: articles/2016-10-12-functor-adts.txt (hunk 1 above)
Comment (anatoliykmetyuk) [reply articles/2016-10-12-functor-adts.txt]: > Oftentimes, a problem arises that in the base code{trait} we do not know the child type the function is called on

I don't think we don't know the _child_ type. We don't know the return type of the child function. Thing is, sometimes `this.type` is not feasible solution. It is feasible only in cases like:

``` scala
trait Bar {
  override def f: Bar
}

trait Foo extends Bar {
  override def f: Foo
}
```

But if we have:

``` scala
trait Bar {
  override def f: CharSequence
}

trait Foo extends Bar {
  override def f: String
}
```

`this.type` is not going to work.

Comment (tindzk) [reply articles/2016-10-12-functor-adts.txt]: I see, fixed. Can I merge the pull request?

**Pull request merged**